agent/exec/containerd/executor.go

```diff
@@ -0,0 +1,127 @@
+package containerd
+
+import (
+	"net"
+	"os"
+	"path/filepath"
+	"runtime"
+	"time"
+
+	"github.com/docker/docker/pkg/sysinfo"
+	"github.com/docker/docker/pkg/system"
+	"github.com/docker/swarmkit/agent/exec"
+	"github.com/docker/swarmkit/agent/secrets"
+	"github.com/docker/swarmkit/api"
+	"github.com/docker/swarmkit/log"
+	"github.com/pkg/errors"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/grpclog"
+)
+
+type executor struct {
+	conn         *grpc.ClientConn
+	secrets      exec.SecretsManager
+	containerDir string
+}
+
+var _ exec.Executor = &executor{}
+
+func getGRPCConnection(ctx context.Context, sock string) (*grpc.ClientConn, error) {
+	grpclog.SetLogger(log.G(ctx))
+
+	dialOpts := []grpc.DialOption{grpc.WithInsecure(), grpc.WithTimeout(100 * time.Second)}
+	dialOpts = append(dialOpts,
+		grpc.WithDialer(func(addr string, timeout time.Duration) (net.Conn, error) {
+			return net.DialTimeout("unix", sock, timeout)
+		},
+		))
+
+	conn, err := grpc.Dial("unix://"+sock, dialOpts...)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to dial %q", sock)
+	}
+
+	return conn, nil
+}
+
+// NewExecutor returns an executor using the given containerd control socket
+func NewExecutor(sock, stateDir string) (exec.Executor, error) {
+	ctx := log.WithModule(context.Background(), "containerd")
+
+	conn, err := getGRPCConnection(ctx, sock)
+	if err != nil {
+		return nil, err
+	}
+
+	containerDir := filepath.Join(stateDir, "containers")
+
+	return &executor{
+		conn:         conn,
+		secrets:      secrets.NewManager(),
+		containerDir: containerDir,
+	}, nil
+}
+
+// Describe returns the underlying node description from containerd
+func (e *executor) Describe(ctx context.Context) (*api.NodeDescription, error) {
+	ctx = log.WithModule(ctx, "containerd")
+
+	hostname := ""
+	if hn, err := os.Hostname(); err != nil {
```

Comment:
This seems to assume that containerd is running on the same host as this executor. I'm not sure this assumption is appropriate, but then again, there probably isn't any other way to get this information, and it's hard to think of a use case where one would prefer to expose the containerd API on a TCP socket and run the executor on a different machine (is that even possible)? What do you think?